### ISSUE TITLE
userdb: 20-systemd-userdb.conf as default can break existing sshd configurations

### ISSUE THREAD

**Issue by joerggollnick** [systemd/systemd#33648]
### systemd version the issue has been seen with

256.1

### Used distribution

ArchLinux

### Linux kernel version used

6.9.7

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

- Installation of systemd version does not break existing configurations of other software
- in this case sshd allows login with a public key fetched by AuthorizedKeysCommand

### Unexpected behaviour you saw

ssh Too many authentication failures

### Steps to reproduce the problem

1. Add AuthorizedKeysCommand to /etc/ssh/sshd_config
2. Install systemd 256.1 (with https://github.com/systemd/systemd/pull/30061)
3. Run sshd -T | grep -i AuthorizedKeysCommand
4. Try to login with a public key fetched from AuthorizedKeysCommand

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
# sshd -T | grep -i AuthorizedKeysCommand
authorizedkeyscommand /usr/bin/userdbctl ssh-authorized-keys %u
authorizedkeyscommanduser root
```

**Comment by YHNdnzj** [comment]
Note that `userdbctl` supports chainloading other commands with `--chain` switch.

**Comment by joerggollnick** [comment]
To summarize:
- sshd_config Include /etc/ssh/sshd_config.d/*.conf was added a few years ago 
- Only one AuthorizedKeysCommand directive is allowed
- Usual pattern for me was to add AuthorizedKeysCommand at the end of sshd_config
- sshd_config later definitions do *not* override earlier definitions
- systemd adds per default /etc/ssh/sshd_config.d/20-systemd-userdb.conf 
- unexpected manual intervention in openssh packages is required (either create 19-custom-authkeyscommand.conf or userdbctl --chain)
- reboot can cause an unreachable system
- userdbctl --chain has to change user from root to some other user

**Comment by donnieelmore** [comment]
Is there a workaround? 

On the newly release CentOS 10 Stream I can't get SSHD to start after kickstart at all because of `/etc/ssh/sshd_config.d/20-systemd-userdb.conf` symlinking to `/usr/lib/systemd/sshd_config.d/20-systemd-userdb.conf` which does not exist on the filesystem.

**Comment by NekkoDroid** [comment]
@donnieelmore you can mask the specific config file by creating an empty one with the same name in  `/etc/tmpfiles.d/`. Then you just need to remove the existing symlinks and you should be fine.

Not ideal IMO, but that is currently I think the best way to work around this.

**Comment by asauber** [comment]
I also lost access to one of my machines. I use use `authorizedkeyscommand` in the `/etc/ssh/sshd_config` file. This change https://github.com/systemd/systemd/pull/30061 overwrote that setting, and I could no longer access my machine except by virutal serial console, which allowed me to remove the file.

Something that overwrites `authorizedkeyscommand`, at first glance, appears malicious. Glad to see that it wasn't in this case. I would vote for this configuration file to be opt-in.

**Comment by infinitydev** [comment]
I was unable to login via SSH after systemd upgrade as well due to my `AuthorizedKeysCommand` being overridden by the new config file and had to recover manually.

I think the symlink `/etc/ssh/sshd_config.d/20-systemd-userdb.conf` should not be created by default because:

1.  It breaks existing configurations and renders users unable to login as demonstrated by this issue.
2.  A sysadmin should always be aware of which keys are allowed to login. By silently allowing `userdbctl` to provide SSH keys, a new security risk is created by pulling keys from locations the admin is not aware of. Prior to this issue, I had no idea what `userdbctl` is or that systemd introduced JSON user records which apparently can contain authorized SSH keys. I am sure I'm not the only one not aware of this. I have no intention to use that facility, making that config file useless at best and dangerous at worst.

My suggestion is to not create the symlink in `sshd_config.d` by default, but document it somewhere for those who intentionally want to allow SSH keys provided by `userdbctl`.

**Comment by keszybz** [comment]
> On the newly release CentOS 10 Stream I can't get SSHD to start after kickstart at all because of `/etc/ssh/sshd_config.d/20-systemd-userdb.conf` symlinking to `/usr/lib/systemd/sshd_config.d/20-systemd-userdb.conf` which does not exist on the filesystem.

That sounds like a packaging bug. The package should not install the symlink without the file.

**Comment by poettering** [comment]
I don't see how this is actionable to us. There's no way in sshd currently to provide two commands, and we cannot know where the user has added their own script, i.e. before or after our drop-in.

I think the right fix would be for sshd to simply allow multiple AuthorizedKeysCommand, but sshd is not really a particularly responsive upstream I guess, so I doubt we'll get that anytime soon.

userdbctl – as mentioned – already supports `--chain` so that we can chainload another tool once it is done. Distributions that for example want to enable sssd's tool for this, can simply change the command line in AuthorizedKeysCommand to invoke userdb and then chain in sssd's tool via --chain, but that's something for the distros to make, and it would only make sure sssds plugin would work (i.e. exactly one other such command), not everyone elses.

ideally, things like sssd would actually just provide their user records in a fashion compatible with userbd, so that things could just work. But I am not expecting any wonders there. 

Hence, for now downstreams have to decide right now if they either want ssh logins to work out of the box for homed accounts (which is right now the pimary provider of userdb records with SSH data), or for something else. If downstreams don't want to support userdb/homed then they can decide not to package the drop-in hence. But of course, this might some people happy but other unhappy.

An option we could discuss is that with extend the --chain logic in userdbctl to invoke all binaries symlinked into some special directories in a similar fashion as itself was invoked. Then, if userdbctl is hooked as the primary AuthorizedKeysCommand, other packages could hook in their tools via symlinking them into those dirs. But of course, this would not provide magic compatibility for already installed systems with manually configured AuthorizedKeysCommand.

**Comment by asauber** [comment]
> That sounds like a packaging bug. The package should not install the symlink without the file.

The issue is not whether or not the symlink is installed correctly.

The issue is that this feature installs an sshd configuration file that overwrites a user-specified value of AuthorizedKeysCommand, locking the user out of ssh connectivity.

**Comment by poettering** [comment]
One question: my assumption is that the primary other tool that people have in AuthorizedKeysCommand is the one provided by sssd. Is there any other relevant package that installs one? (yes, some specific users might install their own homegrown script there, i am not too interested in that, i am curious about the big picture, i.e about relevant other subsystems that are in conflict here)

**Comment by keszybz** [comment]
> > That sounds like a packaging bug. The package should not install the symlink without the file.
> 
> The issue is not whether or not the symlink is installed correctly.

It's totally possible to have two related issued reported in the same bug. This is the case here.

We have the two unrelated issues: the originally reported issue about the setting for `AuthorizedKeysCommand` replacing a previous value, and the other issue which I was responding to, which is about a dangling symlink, which seems to be a packaging bug (or possibly some user error). Not our bug and not related to the original report.

**Comment by asauber** [comment]
Another issue here is users who are not using AuthorizedKeysCommand or sssd at all. 

Users who are familiar with sshd expect that by default the only way that a keypair could be accepted for login is if the file appears in an authorized_keys file. There is no expectation that a hook is available to allow other means to log onto their system.

**Comment by keszybz** [comment]
https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/7/html/linux_domain_identity_authentication_and_policy_guide/openssh-sssd
has the sssd description of use of AuthorizedKeysCommand.

---

I filed https://src.fedoraproject.org/rpms/systemd/pull-request/157 to disable this in Fedora while we figure out a better solution.

**Comment by keszybz** [comment]
> Users who are familiar with sshd expect that by default the only way that a keypair could be accepted for login is if the pubkey appears in an authorized_keys file. There is no expectation that a hook is available to allow other means to log onto their system.

I don't think this is a real issue. userdbd pulls the data from JSON User Records, which are a way to extend the user configuration. This is always done via an administrative action: administrator adds userdb records, or enables a service that provides them, or enables homed to allow home directories to be managed via homed and then users provide a home with some contents…

In a way that is like saying that administrators expect users to be defined in `/etc/passwd`. This *was* true at some point, but we have other mechanism which provide additional definitions for at least 30 years.

**Comment by infinitydev** [comment]
Pulling authorized keys from userdb should have to be explicitly configured by the admin and should not be enabled by default. We are talking about remote logins from the internet. The average user/admin most likely has no idea about systemd user records (I didn't until yesterday), which will inevitably lead to a greater risk of compromised SSH logins if enabled by default. This authentication mechanism is not documented in sshd's man page. If a systemd upgrade quietly adds authentication via systemd user records and an attacker manages to add an own SSH key to a user record, that would be the last place I would look for traces during an investigation if I had no idea about this mechanism in the first place.

`AuthorizedKeysCommand` probably is not the right way to integrate user records into sshd as it can only be used once. I think this should be implemented directly into OpenSSH with a configuration option in sshd_config so that it can be enabled/disabled like most other SSH authentication mechanisms (if systemd is used). The `20-systemd-userdb.conf` file could be used as a workaround until then, but the admin should have to create the symlink in `/etc/ssh/sshd_config.d/` manually.

**Comment by jlebon** [comment]
Another user of `AuthorizedKeysCommand` is https://github.com/coreos/ssh-key-dir, which is used by FCOS. And so naturally, we hit this issue too in FCOS rawhide: https://github.com/coreos/fedora-coreos-tracker/issues/1775.

And yes, agree that sshd ideally would support multiple `AuthorizedKeysCommand` directives. Related to "run all scripts in dropins", see also https://github.com/coreos/ssh-key-dir/issues/10 along those lines.

Just to emphasize and clarify something that was briefly mentioned in https://github.com/systemd/systemd/issues/33648#issuecomment-2226920656:

> sshd_config later definitions do not override earlier definitions

Right, I think definitely one issue here is this assumption in #30061:

> We put the drop-in relatively early, so that other drop-ins generally will override this.

This is actually flipped: earlier dropins override later ones. So one improvement here would be to just change the priority from 20- to e.g. 80- or 90-. That would likely lower the blast radius of this change by a lot (at least FCOS wouldn't have been affected).

**Comment by poettering** [comment]
> Pulling authorized keys from userdb should have to be explicitly configured by the admin and should not be enabled by default.

ssh keys must be explicitly added to user records, provided by the user/admins. There are no ssh keys silently added, hence maybe don't make such a fuss about this.

**Comment by tYYGH** [comment]
Let me try to explain how this bug feels for the “average” user.

I have read the whole thread, and I understand that this new drop-in is there to help professional use-cases, or at least “serious” systems, in contrast to the average desktop system. This is legitimate and I want to accept that.

I’ve been using a Linux desktop since <2000, but I’m forced to use Windows at work. So I don’t have much opportunity to dig into Linux novelties.
I do have a home server with Archlinux. Nothing fancy, but I try to follow best-practice thanks to all the awesome documentation. But once something is setup (SMTP, XMPP, SSH…), I rarely come back to it, unless it breaks during a system upgrade.

And it broke: my Ansible playbook could not connect to the server any more, even though there is an `authorized_keys` file that permits it. I had never heard of `userdbctl`, nor of SSH’s `AuthorizedKeysCommand`.
So far, I just remove the offending file, but it keeps reappearing whenever I update the system.

You say: “_userdbctl supports `--chain`_”. OK. But I’ve read the [man page](https://www.freedesktop.org/software/systemd/man/latest/userdbctl.html) and I have no idea what the expected output is; I could read the manual for `AuthorizedKeysCommand`, I guess…

You say: “_[…] expect users to be defined in `/etc/passwd`. This was true at some point, but we have other mechanism_”. There is one big difference with the current situation:
* if you define a user in `/etc/passwd`, this user works, and is recognized in the system;
* in contrast, I have a perfectly valid `authorized_keys` file, and it does not work.

So I thought: “_how about declaring my ‘authorized key’ in a user **record**_”; this record thing looks nice, after all. But how?…

**Comment by poettering** [comment]
@tYYGH there must be some misunderstanding: AuthorizedKeysCommand is tried *in* *addition* *to* AuthorizedKeysFile by sshd. or in other words, intrepretation of your authorized_keys file is independent of the issue discussed here, and must be caused by something else.

Please do not muddy the waters here.

**Comment by poettering** [comment]
btw, as i understand there are exactly two relevant other implementations around for AuthorizedKeysCommand, both from the general direction of RH, and probably not much used elsewhere: sssd + fcos. 

Which makes me wonder, given this is from the same general direction, has there been any work been done, thoughts spent on figuring out how to marry that sssd and that fcos thing?

also, let's say systemd's userdbctl by default would invoke all binaries symlinked into let's say `/usr/lib/systemd/ssh-authorized-keys-command.d/`, would sssd/fcos people be willing to make use of that?

am i correct in the assumption that both sssd's and fcos' implementation of this hook expect to be called as root? because if so, this would be very compatible with chainloading from userdbctl.

this would then mean that sssd/fcos could still drop in their own sshd_config.d/ file in a way that it is overriden if systemd-userdbd is installed. That would then mean: fcos alone, sssd alone, userdbd alone would all work fine, as would fcos+userdbd, sssd+userdbd and fcos+sssd+userdbd.  But what wouldn#t work of course is just fcos+ssdd, because they would kick out each other with nothing dispatching to both. but maybe that's ok? (and also, i personally wouldn't have to care?)

hence, by adding that callout dir to the mix, can we achieve *pax* *ssh-authorized-keys-command*?

**Comment by jlebon** [comment]
> Which makes me wonder, given this is from the same general direction, has there been any work been done, thoughts spent on figuring out how to marry that sssd and that fcos thing?

None that I'm aware of.

> also, let's say systemd's userdbctl by default would invoke all binaries symlinked into let's say `/usr/lib/systemd/ssh-authorized-keys-command.d/`, would sssd/fcos people be willing to make use of that?

I can't speak for the sssd maintainers, but on the ssh-key-dir side, it would probably be OK.

Though this is just layering yet more complexity on top of what is a seemingly simple sshd RFE. (And in fact, https://github.com/coreos/ssh-key-dir itself is just working around sshd not supporting an `authorized_keys.d/` directory.)

**Comment by poettering** [comment]
> Though this is just layering yet more complexity on top of what is a seemingly simple sshd RFE. (And in fact, https://github.com/coreos/ssh-key-dir itself is just working around sshd not supporting an `authorized_keys.d/` directory.)

That is true. But with openssh upstream "simple RFEs" are not "simple". They are not paritcular in love with that pesky Linux thing.

I mean, try your luck!

**Comment by jlebon** [comment]
> (And in fact, [coreos/ssh-key-dir](https://github.com/coreos/ssh-key-dir?rgh-link-date=2024-09-13T17%3A00%3A13Z) itself is just working around sshd not supporting an `authorized_keys.d/` directory.)

Following up on this, looks like this is happening: https://github.com/coreos/ssh-key-dir/issues/188. This would essentially mean that we would no longer have to have an `AuthorizedKeysCommand=`, leaving space for something else to take it. That said, until OpenSSH also adds support for multiple `AuthorizedKeysCommand=` directives, an `ssh-authorized-keys-command.d/`-like directory probably still makes sense.

**Comment by jinnatar** [comment]
To the posed question of "Are there other relevant implementations", I'll add that [Kanidm](https://kanidm.com/) has a kanidm-unixd service which through the use of the `AuthorizedKeysCommand` mechanism implements SSH key distribution. Similarly with that system in place, "seemingly out of nowhere" a host may stop authenticating as configured if installing `systemd-userdbd` because `20-systemd-userdb.conf` silently overrides the config directives.

Whether it's triggered or not depends on how exactly the admin has added it to their configs since Kanidm does not ship a default-enabled SSH config drop-in and asks the admin to add one if they want the feature. (Not intending to take jabs at systemd with that, just different choices leading to different results, both methods have pros and cons.)
